Lib/test/test_functools.py

```diff
@@ -2147,6 +2147,22 @@ def __eq__(self, other):
                 return self.arg == other
         self.assertEqual(i("str"), "str")
 
+    def test_method_register(self):
+        class A:
+            @functools.singledispatch
+            def t(self, arg):
+                return "base"
+            @t.register(int)
+            def _(self, arg):
+                return "int"
+            @t.register(str)
+            def _(self, arg):
+                return "str"
+        a = A()
+        self.assertEqual(a.t(0), "int")
+        self.assertEqual(a.t(''), "str")
+        self.assertEqual(a.t(0.0), "base")
+
```

Comment:
Please add more tests. At least cover these cases:
* ``@classmethod`` and ``@staticmethod``
* method with zero args (it should fail gracefully)
* calling via instance vs calling via class for all of the above

Reply:
`@classmethod` and `@staticmethod` return descriptors, and singledispatch requires a callable, thus these are not compatible with singledispatch (or would require more significant changes than I'm intending). Generally, you put these around the decorator, but for singledispatch that won't work, as that would break `register()`.

Reply:
There should be at least one way to do it. Allowing to use ``singledispatch`` only with instance methods looks like an incomplete solution.